### PR TITLE
Refactor paging strategy and add tests

### PR DESCRIPTION
**Pull request by dmfenton**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -164,121 +166,101 @@ FeatureService.prototype.getObjectIdField = function (info) {
...
+  // TODO make these calls simultaneous
+  this.layerInfo(function (err, layer) {
+    if (err) return callback(new Error(err || 'Unable to get layer metadata'))
+    this.featureCount(function (err, json) {
+      if (err) return callback(err)
+      if (json.count < 1) return callback(new Error('Service returned count of 0'))
+      var oid = this.getObjectIdField(layer)
+      var size = layer.maxRecordCount
+      // TODO flatten this
+      var metadata = {count: json.count, layer: layer, oid: oid, size: size}
```
Comment (ungoldman) [inline index.js]: what do you mean by flatten?

Comment (dmfenton) [inline index.js]: merge the objects returned into one

File: index.js [hunk 2]
```diff
@@ -164,121 +166,101 @@ FeatureService.prototype.getObjectIdField = function (info) {
  */
 FeatureService.prototype.layerIds = function (callback) {
   this.request(this.url + '/' + this.layer + '/query?where=1=1&returnIdsOnly=true&f=json', function (err, json) {
+    if (!json.objectIds) return callback(new Error('Request for object ids failed'))
+    // TODO: is this really necessary
+    json.objectIds.sort(function (a, b) { return a - b })
     callback(err, json.objectIds)
```
Comment (ungoldman) [inline index.js:172]: should catch and return `err` first then if all's well end with `callback(null, json.objectIds)`

File: index.js [hunk 3]
```diff
@@ -164,121 +166,101 @@ FeatureService.prototype.getObjectIdField = function (info) {
...
-          // default to sequential objectID paging starting from zero
-          return callback(null, this._objectIdPages(0, count, maxCount))
-        }
-      }
-    }.bind(this))
+      }.bind(this))
+    } else {
+      // this is the last thing we can try
+      this.layerIds(function (err, ids) {
+        callback(err, this._idPages(ids, size))
```
Comment (ungoldman) [inline index.js]: same as above

File: index.js [hunk 4]
```diff
@@ -164,121 +166,101 @@ FeatureService.prototype.getObjectIdField = function (info) {
...
+
+    // if the service supports statistics, we can request the maximum and minimum id to build pages
+    if (layer.supportsStatistics) {
+      this._getIdRangeFromStats(meta, function (err, stats) {
+      // if this worked then we can pagination using where clauses
+        if (!err) return callback(null, this._rangePages(stats, size))
+        // if it failed, try to request all the ids and split them into pages
+        this.layerIds(function (err, ids) {
+          // either this works or we give up
+          return callback(err, this._idPages(ids, size))
```
Comment (ungoldman) [inline index.js]: same as above

**Comment by ungoldman** [comment]
Looks good other than minor stuff from comments above. I'm not noticing any breaking changes in external API, are there any that I didn't catch?

File: index.js (hunk 1 above)
Comment (ungoldman) [inline index.js]: ah ok

File: index.js (hunk 3 above)
Comment (dmfenton) [inline index.js]: no need for another line here, this will callback with the error if it fails.

File: index.js (hunk 4 above)
Comment (dmfenton) [inline index.js]: same as above, this will call back with the error if it fails. no need to be verbose

**Comment by dmfenton** [comment]
Thanks for the review. There are no breaking changes in the API. We should release this as a minor version bump.

File: index.js (hunk 3 above)
Comment (ungoldman) [inline index.js]: Right, but if there is an error, I would guess it's not a great idea to still execute `this._idPages(ids, size)`.

File: index.js (hunk 4 above)
Comment (ungoldman) [inline index.js]: same as above. `err` implies `ids` should be null, so `this._idPages` shouldn't be called.

**Comment by dmfenton** [comment]
Ok @ngoldman you win :)

**Comment by ungoldman** [comment]
w00t

**Pull request merged**